haskell/private/haskell_impl.bzl

```diff
@@ -53,6 +53,20 @@ load("@rules_cc//cc:find_cc_toolchain.bzl", "find_cc_toolchain")
 load("//haskell/experimental:providers.bzl", "HaskellModuleInfo")
 load("//haskell/experimental/private:module.bzl", "build_haskell_modules", "get_module_path_from_target")
 
+# Note [Empty Libraries]
+#
+# GHC 8.10.x wants to load the shared libraries corresponding to packages needed
+# for running TemplateHaskell splices. It wants to do this even when all the
+# necessary object files are passed in the command line.
+#
+# In order to satisfy GHC, and yet avoid passing the linked library as input, we
+# create a ficticious package which points to an empty shared library. The
+# ficticious and the real package share the same interface files.
+#
+# Avoiding to pass the real shared library as input is necessary when building
+# individual modules with haskell_module, otherwise building the module would
+# need to wait until all of the modules of library dependencies have been built.
```

Comment:
Thank you for adding this!

Do we also document the motivation in a note somewhere? If not, I think that would be good to document as well[ either here, or in a separate note.

Reply:
> Do we also document the motivation in a note somewhere?

I'm sorry, I think I need this spelled more explicitly. Which motivation would you like to see in a note?

Reply:
The narrowed inputs. Essentially, why is the following a problem that we want to solve?

> GHC 8.10.x wants to load the shared libraries corresponding to packages needed for running TemplateHaskell splices

Reply:
Does the last paragraph in the note address that question?
```
# Avoiding to pass the real shared library as input is necessary when building
# individual modules with haskell_module, otherwise building the module would
# need to wait until all of the modules of library dependencies have been built.
```
The other way to go about it would be to write a grand story for what narrowing does and how it works. This would amount roughly to putting the description of #1631 in a comment and providing links to the intervening parts of the implementation. GHC has some notes like that.

Reply:
> The other way to go about it would be to write a grand story for what narrowing does and how it works.

I tried a note in 02778b9.

Reply:
Thank you! Yes, I think that describes it well.